_posts/2018-09-06-fullstack-typed-api-with-typescript.md

```diff
@@ -0,0 +1,427 @@
+            return new Promise(resolve => resolve(filteredCustomers))
+        }
+    }
+};
+
+// ...
+
+const app = express();
+app.get("/customers", (req, res) => {
+    api.customers.GET().then(msg => {
+        res.send({message: msg});
+    });
+});
+app.post("/customers", (req, res) => {
+    api.customers.POST(req.body.message).then(msg => {
+        res.send({message: msg});
+    });
+});
+app.get("/search", (req, res) => {
+    api.search.GET(req.query).then(msg => {
+        res.send({message: msg});
+    });
+});
+```
+
+If the client tries to call the API with the wrong parameter...
+
+```ts
+api.customers.POST(123);
+```
+
+... the compiler notices this:
+
+```
+error TS2345: Argument of type '123' is not assignable to parameter of type 'string'.
+```
+
+However, we have a problem. The implementations are not completely type checked
+to match the API! In fact, there is a small typo in the client side
+implementation - can you spot it? In addition, there's a lot of duplicated code
+we'd like to avoid. The approach seems to be a good one, but the actual
+implementation needs some work.
+
+Instead of being a type, what if the API was a value? Then we could
+automatically generate code for client fetches and server binds by iterating
+over the API. Turns out, we can, and we also get to maintain type safety.
+
+## Make the object out of you... er, the API
+
+Let's think for a moment. What do we actually want? We want for
+``customers.GET`` to be a function ``() => Promise<Customer[]>`` that is
+located in ``api.customers.GET``. Let's make helper function ``zero<T>(returns:
+T): (T) => Promise<T>`` which, when called with ``zero<Customer[]>([{id: 0,
+name: ''}])``, returns something which has the type ``() =>
+Promise<Customer>``.
+
+```ts
+function zero<R>(returns: R) {
+    return undefined as any as () => Promise<R>;
+}
+```
+...and do the same for ``one()``, which, when called with ``one<string,
+Customer>('', {id: 0, name: ''})``, returns something which has the type ``(p:
+string) => Promise<Customer>``.
+
+```ts
+function one<T, R>() {
+    return undefined as any as (takes: T) => Promise<R>;
+}
+```
+
+Using the new syntax introduced in TypeScript 3.0, we can combine these
+functions into one, and also provide a version, which takes two parameters:
+
+```ts
+// the first any? here is body type, the second is query parameters
+function fun<T, TS extends [any?, any?]>(returns: T, ...takes: TS) {
+    return undefined as any as (...t: TS) => Promise<T>;
+}
+```
+
+Let's also create helpers for string, number, object and array in the similar
+fashion:
+
+```ts
+const str = undefined as any as string;
+const num = undefined as any as string;
+
+function optional<T>(param: T): T | undefined {
+    return undefined as any;
+}
+
+function obj<T extends object>(param: T): T {
+    return undefined as any;
+}
+
+function arr<T>(param: T): T[] {
+    return undefined as any;
+}
+```
+
+We can now write the API in a more terse syntax:
```

Comment:
Maybe the syntax is also more terse, but maybe something like "more succinctly" would be more apt?